src/circuit/server/index.ts

```diff
@@ -0,0 +1,427 @@
+
+    const hopTimeoutPromise = pDefer<HopMessage>()
+    const timeout = setTimeout(() => {
+      hopTimeoutPromise.reject('timed out')
+    }, this.hopTimeout)
+    const pbstr = pbStream(stream)
+
+    try {
+      const request: HopMessage = await Promise.race([
+        pbstr.pb(HopMessage).read(),
+        hopTimeoutPromise.promise
+      ])
+
+      if (request?.type == null) {
+        throw new Error('request was invalid, could not read from stream')
+      }
+
+      log('received', request.type)
+
+      await Promise.race([
+        this.handleHopProtocol({
+          connection,
+          stream: pbstr,
+          request
+        }),
+        hopTimeoutPromise.promise
+      ])
+    } catch (err: any) {
+      log.error('error while handling hop', err)
+      pbstr.pb(HopMessage).write({
+        type: HopMessage.Type.STATUS,
+        status: Status.MALFORMED_MESSAGE
+      })
+      stream.abort(err)
+    } finally {
+      clearTimeout(timeout)
+    }
+  }
+
+  async handleHopProtocol ({ stream, request, connection }: HopProtocolOptions): Promise<void> {
+    log('received hop message')
+    switch (request.type) {
+      case HopMessage.Type.RESERVE: await this.handleReserve({ stream, request, connection }); break
+      case HopMessage.Type.CONNECT: await this.handleConnect({ stream, request, connection }); break
+      default: {
+        log.error('invalid hop request type %s via peer %s', request.type, connection.remotePeer)
+        stream.pb(HopMessage).write({ type: HopMessage.Type.STATUS, status: Status.UNEXPECTED_MESSAGE })
+      }
+    }
+  }
+
+  async handleReserve ({ stream, request, connection }: HopProtocolOptions): Promise<void> {
+    const hopstr = stream.pb(HopMessage)
+    log('hop reserve request from %s', connection.remotePeer)
+
+    if (isRelayAddr(connection.remoteAddr)) {
+      log.error('relay reservation over circuit connection denied for peer: %p', connection.remotePeer)
+      hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+      return
+    }
+
+    if ((await this.connectionGater.denyInboundRelayReservation?.(connection.remotePeer)) === true) {
+      log.error('reservation for %p denied by connection gater', connection.remotePeer)
+      hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+      return
+    }
+
+    const result = await this.reservationStore.reserve(connection.remotePeer, connection.remoteAddr)
+
+    if (result.status !== Status.OK) {
+      hopstr.write({ type: HopMessage.Type.STATUS, status: result.status })
+      return
+    }
+
+    try {
+      // tag relay target peer
+      // result.expire is non-null if `ReservationStore.reserve` returns with status == OK
+      if (result.expire != null) {
+        const ttl = new Date().getTime() - result.expire
+        await this.peerStore.tagPeer(connection.remotePeer, RELAY_SOURCE_TAG, { value: 1, ttl })
```

Comment:
what does `value: 1` mean?

Reply:
The connection manager has a max connections setting - when a node breaches this setting the connection manager chooses connections to prune.

It does this by sorting the connections by peer value and peer values are derived from their tags.

Random incoming dials will have a value of 0 so will be pruned first - setting the tag value to 1 here makes the peer ever so slightly more valuable than a random incoming dial but not as valuable as a peer that is kad-close, a bootstrap node during startup or a in a gossipsub topic mesh, for example.